runtime/vm/ValueTypeHelpers.cpp

```diff
@@ -28,34 +28,55 @@
+		} else {
+			entry->offset = instanceFieldOffset(currentThread, clazz, J9UTF8_DATA(name), J9UTF8_LENGTH(name), J9UTF8_DATA(signature), J9UTF8_LENGTH(signature), &definingClass, (UDATA *)&field, 0);
+		}
+		Assert_VM_unequal(UDATA_MAX, entry->offset);
+	}
+}
+
 void
 defaultValueWithUnflattenedFlattenables(J9VMThread *currentThread, J9Class *clazz, j9object_t instance)
 {
-        J9FlattenedClassCacheEntry * entry = NULL;
-        J9Class * entryClazz = NULL;
-        UDATA length = clazz->flattenedClassCache->numberOfEntries;
-        UDATA const objectHeaderSize = J9VMTHREAD_OBJECT_HEADER_SIZE(currentThread);
-        for (UDATA index = 0; index < length; index++) {
-                entry = J9_VM_FCC_ENTRY_FROM_CLASS(clazz, index);
-                entryClazz = entry->clazz;
-                if (J9_ARE_NO_BITS_SET(J9ClassIsFlattened, entryClazz->classFlags)) {
-                        if (entry->offset == UDATA_MAX) {
-                                J9Class *definingClass = NULL;
-                                J9ROMFieldShape *field = NULL;
-                                J9ROMFieldShape *entryField = entry->field;
-                                J9UTF8 *name = J9ROMFIELDSHAPE_NAME(entryField);
-                                J9UTF8 *signature = J9ROMFIELDSHAPE_SIGNATURE(entryField);
-                                entry->offset = instanceFieldOffset(currentThread, clazz, J9UTF8_DATA(name), J9UTF8_LENGTH(name), J9UTF8_DATA(signature), J9UTF8_LENGTH(signature), &definingClass, (UDATA *)&field, 0);
-                                Assert_VM_notNull(field);
-                        }
-                        MM_ObjectAccessBarrierAPI objectAccessBarrier(currentThread);
-                        objectAccessBarrier.inlineMixedObjectStoreObject(currentThread, 
-                                                                                instance, 
-                                                                                entry->offset + objectHeaderSize, 
-                                                                                entryClazz->flattenedClassCache->defaultValue, 
-                                                                                false);
-                }
-        }
+	J9FlattenedClassCacheEntry * entry = NULL;
+	J9Class * entryClazz = NULL;
+	UDATA length = clazz->flattenedClassCache->numberOfEntries;
+	UDATA const objectHeaderSize = J9VMTHREAD_OBJECT_HEADER_SIZE(currentThread);
+	for (UDATA index = 0; index < length; index++) {
+			entry = J9_VM_FCC_ENTRY_FROM_CLASS(clazz, index);
+			entryClazz = J9_VM_FCC_CLASS_FROM_ENTRY(entry);
+			if (!J9_VM_FCC_ENTRY_IS_STATIC_FIELD(entry) && J9_ARE_NO_BITS_SET(J9ClassIsFlattened, entryClazz->classFlags)) {
+				MM_ObjectAccessBarrierAPI objectAccessBarrier(currentThread);
```

Comment:
If you think there will be multiple static fields, might be worth moving this out of the loop.